keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+Key Hierarchy in KMS plugin (reference implementation):
+
+1. No changes to the API server, keep 1:1 DEK mapping
+    1. Assumption: A KMS plugin that was implemented using a local HSM would not need any changes because it would be able to handle the amount of encryption calls with ease since it would not need to perform network IO
+    1. Assumption: local gRPC calls to the KMS plugin do not represent significant overhead
+1. KMS plugin generates its own local KEK in-memory
+1. External KMS is used to encrypt the local KEK
+1. Local KEK is used for encryption of DEKs sent by API server
+1. Local KEK is used for encryption based on policy (N events, X time, etc)
+
+Since key hierarchy is implemented at the KMS plugin level, it should be seamless for the kube-apiserver. So whether the plugin is using a key hierarchy or not, the kube-apiserver should behave the same.
+
+What is required of the kube-apiserver is to be able to tell the KMS plugin which KEK (local KEK or KMS KEK) it should use to decrypt the incoming DEK. To do so, upon encryption, the KMS plugin could provide the encrypted local KEK as part of the `metadata` field in the `EncryptResponse`. The kube-apiserver would then store it in etcd next to the DEK. Upon decryption, the kube-apiserver provides the encrypted local KEK in `metadata` and `observedKeyID` from the last encryption when calling Decrypt. In case no encrypted local KEK is provided in the `metadata`, then we can assume key hierarchy is not used. The KMS plugin would query the external KMS to use the remote KEK to decrypt the DEK (same behavior as today). No state coordination is required between different instances of the KMS plugin.
+
+For the reference KMS plugin, the encrypted local KEK is stored in etcd via the `metadata` field, and once decrypted, it can be stored in memory as part of the KMS plugin cache to be used for encryption and decryption of DEKs. The encrypted local KEK is used as the key and the decrypted local KEK is stored as the value.
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes ciphertext = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string currentKeyID = 2;
+    // Additional metadata to be stored with the encrypted data.
+    // This metadata can contain the encrypted local KEK that was used to encrypt the DEK.
+    // Stored unencrypted in etcd.
+    map<string, string> metadata = 3;
+}
+```
+
+The `DecryptRequest` passes the same `currentKeyID` and `metadata` returned by the previous `EncryptResponse` of this data as its `observedKeyID` and `metadata` for the decryption request.
+
+```proto
+message DecryptRequest {
+    // The data to be decrypted.
+    bytes ciphertext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
```

Comment:
Should this be request metadata rather than embedding in the request proto?

Reply:
🤷 what is gRPC best practice?

Reply:
I think request metadata would be better than embedding in the request proto. Request metadata is essentially headers:

https://github.com/grpc/grpc-go/blob/master/Documentation/grpc-metadata.md

Reply:
@mikedanese We decided to keep UID as part of the proto since it is a core change we are introducing in v2 and it's not optional. Having it in the proto will make it easily accessible and ensuring it is something that the plugins can rely on.